File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        
        q.offer(root);
        int level = 0;
        
        while (!q.isEmpty()) {
            int size = q.size();
            List<Integer> levelOrder = new ArrayList<>();
            
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                if (node == null) {
                    continue;
                }
                
                levelOrder.add(node.val);
                
                q.offer(node.left);
                q.offer(node.right);
            }
        
            if (level % 2 == 1) {
                Collections.reverse(levelOrder);
            }      
            
            if (!levelOrder.isEmpty()) {
                ans.add(levelOrder);
            }
            
            level++;
        }
        
        return ans;
    }
}

File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        if root is None:
            return res
        
        q = deque()
        q.append(root)
        forward = True
        while q:
            size = len(q)
            next_q = deque()
            level = []
            for _ in range(size):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    next_q.append(node.left)
                if node.right:
                    next_q.append(node.right)
            if forward:
                res.append(level)
            else:
                res.append(reversed(level))
            q = next_q
            forward = not forward
        
        return res